controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/NameServiceDispatcherTest.java

```diff
@@ -0,0 +1,122 @@
+        assertEquals(0.5, dispatcher.maintain()); // 2 of 4 requests were ok (the fourth was never attempted).
+        assertEquals(List.of(req4, req6), tester.curator().readNameServiceQueue().requests());
+
+        // Queue again initially contains a subsequence of requests, 3–6.
+        // While the dispatcher is working through those, some requests are appended, and some are prepended.
+        // The dispatch of requests owned by owner1 will fail, so the original sublist read by the dispatcher
+        // is not removed, but replaced by a subsequence, upon dispatch end.
+        tester.curator().writeNameServiceQueue(new NameServiceQueue(base));
+        expectations.clear();
+        expectations.add(name -> {
+            assertEquals(rec1.name(), name);
+            tester.curator().writeNameServiceQueue(tester.curator().readNameServiceQueue()
+                                                         .with(req7)
+                                                         .with(req2, Priority.high)
+                                                         .with(req1, Priority.high)
+                                                         .with(req8));
+            assertEquals(List.of(req1, req2, req3, req4, req5, req6, req7, req8), tester.curator().readNameServiceQueue().requests());
+        });
+        expectations.add(failOwner1.get()); // Recursively adds itself to the tail of the expectation queue.
+        expectations.add(name -> assertEquals(rec3.name(), name));
+        assertEquals(0.5, dispatcher.maintain()); // 2 of 4 requests were ok (the fourth was never attempted).
+        assertEquals(List.of(req1, req2, req4, req6, req7, req8), tester.curator().readNameServiceQueue().requests());
+
+
+        // Finally, queue again initially contains a subsequence of requests, 3–6.
+        // While the dispatcher is working through those, the queue is altered in unexpected ways—specifically, owner2's requests is gone.
+        // The dispatch of requests owned by owner1 still fail, so the original sublist read by the dispatcher
+        // is not removed, nor is it replaced by a subsequence, but the processed requests are attempted removed from the current queue.
+        tester.curator().writeNameServiceQueue(new NameServiceQueue(base));
+        expectations.clear();
+        expectations.add(name -> {
+            assertEquals(rec1.name(), name);
+            tester.curator().writeNameServiceQueue(new NameServiceQueue(List.of(req1, req2, req3, req4, req6, req7, req8)));
+        });
+        expectations.add(failOwner1.get()); // Recursively adds itself to the tail of the expectation queue.
+        expectations.add(name -> assertEquals(rec3.name(), name));
+        assertEquals(0.5, dispatcher.maintain()); // 2 of 4 requests were ok (the fourth was never attempted).
+        assertEquals(List.of(req2, req3, req4, req6, req7, req8), tester.curator().readNameServiceQueue().requests());
```

Comment:
Note the "error" of removing req1 instead of req3, which is really the same request. If req2 was req1's inverse, this would be significant. The odds of this happening seem _very_ low.